fortanix-vme/nsm/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+            ErrorCode::ReadOnlyIndex => Error::ReadOnlyPcrIndex,
+            ErrorCode::InvalidOperation => Error::InvalidOperation,
+            ErrorCode::BufferTooSmall => Error::BufferTooSmall,
+            ErrorCode::InputTooLarge => Error::InputTooLarge,
+            ErrorCode::InternalError => Error::InternalError,
+            ErrorCode::Success => Error::InvalidResponse,
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct Pcr {
+    locked: bool,
+    data: Vec<u8>,
+}
+
+impl Pcr {
+    fn new(locked: bool, data: Vec<u8>) -> Self {
+        Pcr {
+            locked,
+            data,
+        }
+    }
+
+    pub fn locked(&self) -> bool {
+        self.locked
+    }
+
+    pub fn data(&self) -> &[u8] {
+        self.data.as_slice()
+    }
+}
+
+impl TryFrom<Response> for Pcr {
+    type Error = Error;
+
+    fn try_from(req: Response) -> Result<Self, Self::Error> {
+        match req {
+            Response::DescribePCR { lock, data } => Ok(Pcr::new(lock, data)),
+            Response::ExtendPCR { data }         => Ok(Pcr::new(false, data)) /* Only unlocked PCRs can get extended */,
+            Response::Error(code)                => Err(code.into()),
+            _                                    => Err(Error::InvalidResponse),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Description {
+    /// Breaking API changes are denoted by `major_version`
+    pub version_major: u16,
+    /// Minor API changes are denoted by `minor_version`. Minor versions should be backwards compatible.
+    pub version_minor: u16,
+    /// Patch version. These are security and stability updates and do not affect API.
+    pub version_patch: u16,
+    /// `module_id` is an identifier for a singular NitroSecureModule
+    pub module_id: String,
+    /// The maximum number of PCRs exposed by the NitroSecureModule.
+    pub max_pcrs: u16,
+    /// The PCRs that are read-only.
+    pub locked_pcrs: BTreeSet<u16>,
+    /// The digest of the PCR Bank
+    pub digest: Digest,
+}
+
+impl Description {
+    pub fn version_major(&self) -> u16 {
+        self.version_major
+    }
+
+    /// Minor API changes are denoted by `minor_version`. Minor versions should be backwards compatible.
+    pub fn version_minor(&self) -> u16 {
+        self.version_minor
+    }
+
+    /// Patch version. These are security and stability updates and do not affect API.
+    pub fn version_patch(&self) -> u16 {
+        self.version_patch
+    }
+
+    /// `module_id` is an identifier for a singular NitroSecureModule
+    pub fn module_id(&self) -> &String {
+        &self.module_id
+    }
+
+    /// The maximum number of PCRs exposed by the NitroSecureModule.
+    pub fn max_pcrs(&self) -> u16 {
+        self.max_pcrs
+    }
+
+    /// The PCRs that are read-only.
+    pub fn locked_pcrs(&self) -> &BTreeSet<u16> {
+        &self.locked_pcrs
+    }
+
+    /// The digest of the PCR Bank
+    pub fn digest(&self) -> Digest {
+        self.digest
+    }
```

Comment:
Are these getters needed, since the fields are public?

Reply:
Oops. I created these getters as clients never need to create their own `Description` value. So having one, must have come from the `nsm` module. I forgot to make the fields private afterwards.